### ISSUE TITLE
Frozen-string-literal comment detection does not detect comment headers.

### ISSUE THREAD

**Issue by envygeeks** [bbatsov/rubocop#2643]
A comment header does not affect `Frozen-string-literal: true` in Ruby, however, when using comment headers Rubocop does not properly detect `Frozen-string-literal: true`:

``` ruby
# ----------------------------------------------------------------------------
# Frozen-string-literal: true
# Copyright: 2015 - 2016 Jordon Bedwell - Apache v2.0 License
# Encoding: utf-8
# ----------------------------------------------------------------------------
```

``` ruby
Luna::RSpec::Formatters::Profile::EXAMPLES.frozen? # => true
```

```
lib/luna/rspec/formatters/profiles.rb:1:1: C: Missing frozen string literal comment.
# ----------------------------------------------------------------------------
^
```

**Comment by gylaz** [comment]
Should that be `# frozen_string_literal: true`? Either way, RuboCop doesn't seem to pick up on that comment option either.

**Comment by envygeeks** [comment]
The example I gave clearly indicates that the way I did it works...  Rubo cop does pick up the comment without the header, try using the master branch instead of the stable release.

**Comment by durrantm** [comment]
Yes it should be `frozen_string_literal: true` as `Frozen_string_literal:true` doesn't work

**Comment by envygeeks** [comment]
@durrantm sorry, but you are wrong.
```
cat test1.rb && bundle exec ruby test1.rb && printf "\n\n" && \
cat test2.rb && bundle exec ruby test2.rb

# Frozen-string-literal: true
$stdout.puts "".frozen?
true


# Frozen-STRING-Literal: true
$stdout.puts "".frozen?
true
```

**Comment by hendie** [comment]
I bumped my head against this one today: @envygeeks in the example you show at the top of this thread, you wrote `Frozen-string-literal: true`. Rubocop didn't get that one for me either, but when I changed to underscores: ` Frozen_string_literal: true` everything magically started working (as in the later examples above). Those underscores can be ugly gotchas!

**Comment by mikegee** [comment]
I'm having a hard time finding a definitive answer on the format of Ruby's magic comments, so I struggled through Ruby's `parse.y` to figure it out myself.

I'm pretty sure [this part](https://github.com/ruby/ruby/blob/78b95b49f8715a4782f5b9bdc4c163e445cdc303/parse.y#L7134-L7136) means dashes and underscores are equivalent and [this case-insensitive comparison](https://github.com/ruby/ruby/blob/78b95b49f8715a4782f5b9bdc4c163e445cdc303/parse.y#L7138) means case doesn't matter.

I suspect a minor adjustment to [the regex in `Rubocop::MagicComment`](https://github.com/bbatsov/rubocop/blob/d76dc907dc0f249615aa8a6b003f50247b187747/lib/rubocop/magic_comment.rb#L192) will resolve this issue.

**Comment by backus** [comment]
@mikegee correct that adjusting that regex will fix the issue.

> I'm having a hard time finding a definitive answer on the format of Ruby's magic comments, so I struggled through Ruby's parse.y to figure it out myself.

You are correct that it is barely documented. I include three relevant links in the docs for that `MagicComment` file. Beyond that there is the `parse.y` which you have correctly dug through to find the correct behavior. It looks like changing the separators in that regex to be `[_\-]` and to add the `i` regex option would address this issue.

**Comment by mikegee** [comment]
@envygeeks While looking into this issue I found that the `Encoding` comment must be the first line of code (except an optional shebang line) for Ruby to respect your declaration.

Since UTF-8 is the default source encoding since Ruby 2.0, this issue is moot in the particular snippet you posted.

This can be enforced by the `Style/Encoding` cop by enabling it and setting its config to always expect an encoding comment:

```yaml
Style/Encoding:
  Enabled: true
  EnforcedStyle: always
```

`Style/Encoding` [is disabled by default](https://github.com/bbatsov/rubocop/blob/d76dc907dc0f249615aa8a6b003f50247b187747/config/disabled.yml#L33-L36) and [would enforce never including an encoding comment](https://github.com/bbatsov/rubocop/blob/d76dc907dc0f249615aa8a6b003f50247b187747/config/default.yml#L468-L469) if simply enabled.
